src/db.rs

```diff
@@ -10,22 +10,145 @@
 //! Finding the "rpm-devel" RPM in the database:
 //!
 //! ```
-//! use librpm::Index;
+//! use librpm::{Db, Index};
 //!
-//! librpm::config::read_file(None).unwrap();
+//! use std::path::Path;
 //!
-//! let mut matches = Index::Name.find("rpm-devel");
+//! let db = Db::open::<&Path>().unwrap();
+//! let mut matches = Index::Name.find(&db, "rpm-devel");
 //! let package = matches.next().unwrap();
 //!
 //! println!("package name: {}", package.name);
 //! println!("package summary: {}", package.summary);
 //! println!("package version: {}", package.version);
 //! ```
 
+use crate::error::{Error, ErrorKind};
 use crate::internal::{iterator::MatchIterator, tag::Tag};
 use crate::package::Package;
 use streaming_iterator::StreamingIterator;
 
+use std::cell::Cell;
+use std::ffi::CString;
+use std::fmt;
+use std::marker::PhantomData;
+use std::os::unix::ffi::OsStrExt;
+use std::path::Path;
+use std::ptr;
+
+static mut LIB_RPM_CONFIGURED: bool = false;
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Db {
+    _this_type_is_not_sync: PhantomData<Cell<()>>,
```

Comment:
While this prevents the `Send` case of thread unsafety, it doesn't appear to prevent the caller from instantiating `Db` multiple times. which would still elicit the thread unsafety problems.

I think a better option is a `Mutex`. Originally I had a `Mutex` around the FFI, held by the open `TransactionSet`:

https://github.com/iqlusioninc/crates/pull/28/files#diff-633fedf9a15d1830af2e8851d032cd59L14

Reply:
Multiple instances are prevented via returning `Err` from `open`. I also have a global flag for native library configuration - although it's just a static `bool` for now.